### PR TITLE
Split tests into separate classes

### LINKED ISSUE #1494: Refactor tests into separate classes
Currently the test project has one single class containing all tests.  For better separation of logic amongst the the image types, it would be better to split the tests into separate classes with each class handling one of the image types (runtime, aspnet, runtime-deps, sdk).  This should also help to eliminate a current issue where some SDK tests end up running a test more than once on the same SDK version because of how the ImageData is constructed.

### PR DESCRIPTION
**Pull request by mthalman**
This organizes tests a bit better so that they are split into image type-specific classes.  This also avoids duplication of tests in scenarios where a custom SdkOS is defined by filtering the input to `[MemberData]`.

Fixes #1494

### CODE REVIEW

File: tests/Microsoft.DotNet.Docker.Tests/RuntimeDepsImageTests.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        public static IEnumerable<object[]> GetImageData()
+        {
+            return TestData.GetImageData()
+                .Distinct(new DefaultImageDataEqualityComparer())
+                .Select(imageData => new object[] { imageData });
+        }
+
+        [Theory]
```
Comment (MichaelSimons) [inline tests/Microsoft.DotNet.Docker.Tests/RuntimeDepsImageTests.cs]: There are at least two test cases that could be cleaned up a little if we implemented a `LinuxImageTheory` attribute.  This test as well as `InsecureFiles` could make use of it.

```
public class LinuxImageTheoryAttribute : TheoryAttribute
{
    public LinuxImageTheoryAttribute()
    {
        if (!DockerHelper.IsLinuxContainerModeEnabled)
        {
            Skip = "Linux image test not applicable when running in Windows Container mode";
        }
    }
}
```

File: tests/Microsoft.DotNet.Docker.Tests/CommonImageTests.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            _dockerHelper = new DockerHelper(outputHelper);
+            _outputHelper = outputHelper;
+        }
+
+        public static IEnumerable<object[]> GetImageData()
+        {
+            return TestData.GetImageData()
+                // Filter the image data down to avoid duplicates caused by variations in their SDK OS.
+                .Distinct(new DefaultImageDataEqualityComparer())
```
[the remark this answers is not in the document]
Comment (mthalman) [reply tests/Microsoft.DotNet.Docker.Tests/CommonImageTests.cs]: Because it's using `Distinct`.

[the remark this answers is not in the document]
Comment (MichaelSimons) [reply tests/Microsoft.DotNet.Docker.Tests/CommonImageTests.cs]: What am I missing?  DefaultImageDataEqualityComparer doesn't consider the SdkOS.  The VersionString, OS, and Arch already appear distinct for the ImageData we have.  What does distinct filter out?  

```
            new ImageData { Version = V3_0, OS = OS.Alpine39,     Arch = Arch.Arm64,    SdkOS = OS.Buster },
            new ImageData { Version = V3_0, OS = OS.Alpine310,    Arch = Arch.Arm64,    SdkOS = OS.Buster },
```

Some SDK images will still get tested multiple times.  While this isn't egregious, it isn't optimal and isn't a pattern that bodes well for adding additional "common" tests.

Perhaps base classes would be some help here.  I took your changes and whipped up a quick illustration.  I'd be surprised if it compiled but hopefully you will get the idea.  The main disadvantage is true shared tests have to be declared twice - once in sdk and once in CommonRuntimeImageTests.  https://github.com/dotnet/dotnet-docker/compare/nightly...MichaelSimons:shared-tests?expand=1

File: tests/Microsoft.DotNet.Docker.Tests/SdkImageTests.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                _dockerHelper.Run(
+                    image: imageData.GetImage(DotNetImageType.SDK, _dockerHelper),
+                    command: verifyCacheCommand,
+                    name: imageData.GetIdentifier("PackageCache"));
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(GetImageData))]
+        public void PowerShellScenario_DefaultUser(ImageData imageData)
```
Comment (MichaelSimons) [inline tests/Microsoft.DotNet.Docker.Tests/SdkImageTests.cs]: Consider prefixing with `Verify` for consistency.

File: tests/Microsoft.DotNet.Docker.Tests/CommonImageTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mthalman) [reply tests/Microsoft.DotNet.Docker.Tests/CommonImageTests.cs]: I get it now.  For some reason I was thinking the runtime images had duplicates with only the SDK causing the variation.  I like your inheritance solution.  That actually seems a more natural implementation than having a `CommonImageTests` test class that handles things separately.  Another option would be splitting out `VerifyInsecureFiles` into two separate test methods that make use of different data for `[MemberData]` but that doesn't seem like a natural fit in that class; rather the responsibility of SDK-specific filtering should just get moved to `SdkImageTest`.

File: tests/Microsoft.DotNet.Docker.Tests/CommonRuntimeImageTests.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+        public static IEnumerable<object[]> GetImageData()
+        {
+            return TestData.GetImageData()
+                .Select(imageData => new object[] { imageData });
+        }
+
+        [LinuxImageTheory]
+        [MemberData(nameof(GetImageData))]
+        public void VerifyCommonRuntimeInsecureFiles(ImageData imageData)
```
Comment (MichaelSimons) [inline tests/Microsoft.DotNet.Docker.Tests/CommonRuntimeImageTests.cs]: If you don't like the idea of an `new` or `virtual/override` method here, consider using `VerifyInsecureFiles` for the actual test case name and changing the name of the base method (e.g. `VerifyCommonEnvironmentVariables`).  I think it is nicer to see `Microsoft.DotNet.Docker.Tests.AspnetImageTests.VerifyInsecureFiles` vs `Microsoft.DotNet.Docker.Tests.AspnetImageTests.VerifyCommonRuntimeInsecureFiles` in the AzDO Tests result view.

File: tests/Microsoft.DotNet.Docker.Tests/AspnetImageTests.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        [MemberData(nameof(GetImageData))]
+        public async Task VerifyAppScenario(ImageData imageData)
+        {
+            ImageScenarioVerifier verifier = new ImageScenarioVerifier(imageData, DockerHelper, OutputHelper, isWeb: true);
+            await verifier.Execute();
+        }
+
+        [Theory]
+        [MemberData(nameof(GetImageData))]
+        public void VerifyAspnetEnvironmentVariables(ImageData imageData)
```
[the remark this answers is not in the document]
Comment (mthalman) [reply tests/Microsoft.DotNet.Docker.Tests/AspnetImageTests.cs]: runtime-deps requires the `LinuxImageTheory` attribute which is different than the others which are applicable to both Linux and Windows.

File: tests/Microsoft.DotNet.Docker.Tests/CommonRuntimeImageTests.cs (hunk 4 above)
Comment (mthalman) [reply tests/Microsoft.DotNet.Docker.Tests/CommonRuntimeImageTests.cs]: I'll go with the name change.  There are a variety of xUnit code analyzer warnings that crop up when trying to use `new` or `virtual/override`.

File: tests/Microsoft.DotNet.Docker.Tests/AspnetImageTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply tests/Microsoft.DotNet.Docker.Tests/AspnetImageTests.cs]: Ah, that's unfortunate.  In that case, consider renaming the base method and use `VerifyEnvironmentVariables` for the actual test case names in order to provide a better AzDO experience.

**Pull request merged**